### PR TITLE
doc/md: add doctrine provider guide

### PR DESCRIPTION
**Pull request by ronenlu**

### CODE REVIEW

File: doc/md/guides/orms/doctrine.mdx [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+
+## TL;DR
+* [Doctrine](https://www.doctrine-project.org/) is an ORM library that's widely used in the PHP community.
+* [Atlas](https://atlasgo.io) is an open-source tool for inspecting, planning, linting and
+  executing schema changes to your database.
+* Developers using Doctrine can use Atlas to automatically plan schema migrations
+  for them, based on the desired state of their schema instead of crafting them by hand.
+
+## Automatic migration planning for Doctrine
+Doctrine is a popular ORM widely used in the PHP community, used by projects like Symfony.
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:19]: Link to Symfony

File: doc/md/guides/orms/doctrine.mdx [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+## TL;DR
+* [Doctrine](https://www.doctrine-project.org/) is an ORM library that's widely used in the PHP community.
+* [Atlas](https://atlasgo.io) is an open-source tool for inspecting, planning, linting and
+  executing schema changes to your database.
+* Developers using Doctrine can use Atlas to automatically plan schema migrations
+  for them, based on the desired state of their schema instead of crafting them by hand.
+
+## Automatic migration planning for Doctrine
+Doctrine is a popular ORM widely used in the PHP community, used by projects like Symfony. 
+Doctrine allows users to manage their database schemas using its [orm:schema-tool](https://www.doctrine-project.org/projects/doctrine-orm/en/3.1/reference/tools.html#schematool) Console command,
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:20]:
```suggestion
Doctrine allows users to manage their database schemas using its [orm:schema-tool](https://www.doctrine-project.org/projects/doctrine-orm/en/3.1/reference/tools.html#schematool) Console Command,
```

File: doc/md/guides/orms/doctrine.mdx [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+which is usually sufficient during development and in many simple cases.
+
+However, at some point, teams need more control and decide to employ [versioned migrations](/concepts/declarative-vs-versioned#versioned-migrations) methodology.
+Doctrine has a [Migration tool](https://www.doctrine-project.org/projects/doctrine-migrations/en/3.7/index.html) that can be used to manage versioned migrations.
+Doctrine can also [generate a migration](https://www.doctrine-project.org/projects/doctrine-migrations/en/3.7/reference/generating-migrations.html)
+for you by comparing the current state of your database schema to the state defined by using the ORM.
+
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:27]:
```suggestion
```

File: doc/md/guides/orms/doctrine.mdx [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+
+A downside of this approach is that in order for it to work, a pre-existing database with the current version of the schema must be connected to.
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:29]:
```suggestion
A downside of this approach is that in order for it to work, you must be connected to a pre-existing database with the current version of the schema.
```

File: doc/md/guides/orms/doctrine.mdx [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+In many production environments, databases should generally not be reachable from developer workstations,
+which means this comparison is normally done against a local copy of the database which may have
+undergone some changes that aren't reflected in the existing migrations.
+
+More of the limitations of Doctrine migration tool include:
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:34]:
```suggestion
More of the limitations of the Doctrine migration tool include:
```

File: doc/md/guides/orms/doctrine.mdx [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+
+* **Ensuring Migration Safety.** Migrations are a risky business. If you're not careful, you can easily cause data loss or 
+  a production outage. Doctrine migration tool does not provide a native way to ensure that a migration is safe to apply.
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:37]:
```suggestion
  a production outage. The Doctrine migration tool does not provide a native way to ensure that a migration is safe to apply.
```

File: doc/md/guides/orms/doctrine.mdx [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+* **Modern Deployments.**  Doctrine does not provide native integration with modern deployment practices such as [GitOps](/blog/2023/12/06/gitops-for-databases-part-1)
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:38]:
```suggestion
* **Modern Deployments.**  Doctrine does not provide native integrations with modern deployment practices such as [GitOps](/blog/2023/12/06/gitops-for-databases-part-1)
```

File: doc/md/guides/orms/doctrine.mdx [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+  or [Infrastructure-as-Code](/blog/2023/07/19/bridging-the-gap-between-iac-and-schema-management).
+
+Using Atlas, you can enjoy automatic migration planning, automatic code review and integrations with your favorite CI/CD tools.
+
+Atlas works in similar way to Doctrine migration tool, by calculating the diff between the _current_ state of the database,
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:43]:
```suggestion
Atlas works in similar way to the Doctrine migration tool, by calculating the diff between the _current_ state of the database,
```

File: doc/md/guides/orms/doctrine.mdx [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+</TabItem>
+</Tabs>
+
+### Doctrine Command vs Script mode
+
+The Atlas Doctrine Provider can be used in two modes:
+
+* **Doctrine Command** - If all of your Doctrine entities exist under single directory,
+you can add the atlas-provider command to the [Doctrine Console](https://www.doctrine-project.org/projects/doctrine-orm/en/3.0/reference/tools.html) file.
+* **Script** - In other cases, you can use the provider as a php script to load your Doctrine schema into Atlas.
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:95]:
```suggestion
* **Script** - In other cases, you can use the provider as a PHP script to load your Doctrine schema into Atlas.
```

File: doc/md/guides/orms/doctrine.mdx [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+  INDEX `IDX_1E197C9F8D8CDF1` (`engineer_id`),
+  CONSTRAINT `FK_1E197C9E1CFE6F5` FOREIGN KEY (`reporter_id`) REFERENCES `users` (`id`) ON UPDATE NO ACTION ON DELETE NO ACTION,
+  CONSTRAINT `FK_1E197C9F8D8CDF1` FOREIGN KEY (`engineer_id`) REFERENCES `users` (`id`) ON UPDATE NO ACTION ON DELETE NO ACTION
+) CHARSET utf8mb4 COLLATE utf8mb4_0900_ai_ci;
+
+```
+
+Amazing! Atlas automatically generated a migration file that will create the `users` and `bugs` tables in our database.
+
+Next, alter the `User` entity to include `$name` property.
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:306]:
```suggestion
Next, alter the `User` entity to include a `$name` property.
```

**Review by hilakashai — COMMENTED** [review]
🚀

File: doc/md/guides/orms/doctrine.mdx [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+  executing schema changes to your database.
+* Developers using Doctrine can use Atlas to automatically plan schema migrations
+  for them, based on the desired state of their schema instead of crafting them by hand.
+
+## Automatic migration planning for Doctrine
+Doctrine is a popular ORM widely used in the PHP community, used by projects like [Symfony](https://symfony.com/).
+Doctrine allows users to manage their database schemas using its [orm:schema-tool](https://www.doctrine-project.org/projects/doctrine-orm/en/3.1/reference/tools.html#schematool) Console Command,
+which is usually sufficient during development and in many simple cases.
+
+However, at some point, teams need more control and decide to employ [versioned migrations](/concepts/declarative-vs-versioned#versioned-migrations) methodology.
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:23]:
```suggestion
However, at some point, teams need more control and decide to employ a [versioned migrations](/concepts/declarative-vs-versioned#versioned-migrations) methodology.
```

File: doc/md/guides/orms/doctrine.mdx [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+Doctrine has a [Migration tool](https://www.doctrine-project.org/projects/doctrine-migrations/en/3.7/index.html) that can be used to manage versioned migrations.
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:24]:
```suggestion
Doctrine has a [migration tool](https://www.doctrine-project.org/projects/doctrine-migrations/en/3.7/index.html) that can be used to manage versioned migrations.
```

File: doc/md/guides/orms/doctrine.mdx [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+More of the limitations of the Doctrine migration tool include:
+
+* **Ensuring Migration Safety.** Migrations are a risky business. If you're not careful, you can easily cause data loss or 
+  a production outage. The Doctrine migration tool does not provide a native way to ensure that a migration is safe to apply.
+* **Modern Deployments.**  Doctrine does not provide native integrations with modern deployment practices such as [GitOps](/blog/2023/12/06/gitops-for-databases-part-1)
+  or [Infrastructure-as-Code](/blog/2023/07/19/bridging-the-gap-between-iac-and-schema-management).
+
+Using Atlas, you can enjoy automatic migration planning, automatic code review and integrations with your favorite CI/CD tools.
+
+Atlas works in similar way to the Doctrine migration tool, by calculating the diff between the _current_ state of the database,
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:42]:
```suggestion
Atlas works in a similar way to the Doctrine migration tool, by calculating the diff between the _current_ state of the database,
```

Comment (hilakashai) [reply doc/md/guides/orms/doctrine.mdx:42]: missing the end of this sentence..

File: doc/md/guides/orms/doctrine.mdx [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+
+In the context of [versioned migrations](/concepts/declarative-vs-versioned#versioned-migrations),
+the current state can be thought of as the database schema that would have
+been created by applying all previous migration scripts.
+
+The desired schema of your application can be provided to Atlas via an [External Schema Datasource](/atlas-schema/projects#data-source-external_schema),
+which is any program that can output a SQL schema definition to stdout.
+
+To use Atlas with Doctrine, users can utilize the [Doctrine Atlas Provider](https://github.com/ariga/atlas-provider-doctrine),
+which is a small program that can be used to load the schema of a Doctrine project into Atlas.
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:52]:
```suggestion
which is a small program that loads the schema of a Doctrine project into Atlas.
```

File: doc/md/guides/orms/doctrine.mdx [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+
+In this guide, we will show how Atlas can be used to automatically plan schema migrations for Doctrine users.
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:54]:
```suggestion
In this guide, we will show how to use Atlas to automatically plan schema migrations for Doctrine users.
```

File: doc/md/guides/orms/doctrine.mdx [hunk 16]
```diff
@@ -0,0 +1,343 @@
...
+composer require ariga/atlas-provider-doctrine:^4
+```
+</TabItem>
+</Tabs>
+
+### Doctrine Command vs Script mode
+
+The Atlas Doctrine Provider can be used in two modes:
+
+* **Doctrine Command** - If all of your Doctrine entities exist under single directory,
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:97]:
```suggestion
* **Doctrine Command** - If all of your Doctrine entities exist under a single directory,
```

File: doc/md/guides/orms/doctrine.mdx [hunk 17]
```diff
@@ -0,0 +1,343 @@
...
+you can add the atlas-provider command to the [Doctrine Console](https://www.doctrine-project.org/projects/doctrine-orm/en/3.0/reference/tools.html) file.
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:98]:
```suggestion
you can add the `atlas-provider` command to the [Doctrine Console](https://www.doctrine-project.org/projects/doctrine-orm/en/3.0/reference/tools.html) file.
```

File: doc/md/guides/orms/doctrine.mdx [hunk 18]
```diff
@@ -0,0 +1,343 @@
...
+* **Script** - In other cases, or if you use a [Symfony](https://symfony.com/) project,
+you can use the provider as a PHP script to load your Doctrine schema into Atlas.
+
+### Using the Doctrine Command
+
+Add the atlas-provider command to the Doctrine Console file:
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:104]:
```suggestion
Add the `atlas-provider`command to the Doctrine Console file:
```

File: doc/md/guides/orms/doctrine.mdx [hunk 19]
```diff
@@ -0,0 +1,343 @@
...
+  format {
+    migrate {
+      diff = "{{ sql . \"  \" }}"
+    }
+  }
+}
+```
+
+
+### As PHP Script
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:150]:
```suggestion
### As a PHP Script
```

File: doc/md/guides/orms/doctrine.mdx [hunk 20]
```diff
@@ -0,0 +1,343 @@
...
+
+    #[ORM\ManyToOne(targetEntity: User::class, inversedBy: 'reportedBugs')]
+    private User|null $reporter;
+}
+
+```
+</TabItem>
+</Tabs>
+
+We can generate a migration file by running this command:
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:267]:
```suggestion
We can generate a migration file by running the following command:
```

File: doc/md/guides/orms/doctrine.mdx [hunk 21]
```diff
@@ -0,0 +1,343 @@
...
+
+```bash
+atlas migrate diff --env doctrine
+```
+
+Running this command will generate files similar to this in the `migrations` directory:
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:273]:
```suggestion
Running this command will generate files in the `migrations` directory, similarly to:
```

File: doc/md/guides/orms/doctrine.mdx [hunk 22]
```diff
@@ -0,0 +1,343 @@
...
++   private string $name;
+```
+
+Re-run this command:
+
+```bash
+atlas migrate diff --env doctrine
+```
+
+Observe a new migration file is generated:
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:329]:
```suggestion
Observe that a new migration file is generated:
```

File: doc/md/guides/orms/doctrine.mdx [hunk 23]
```diff
@@ -0,0 +1,343 @@
...
+```sql
+-- Modify "users" table
+ALTER TABLE `users` ADD COLUMN `name` varchar(255) NOT NULL;
+```
+
+## Conclusion
+
+In this guide we demonstrated how projects using Dcotrine can use Atlas to automatically
+plan schema migrations based only on their data model. To learn more about executing
+migrations against your production database, read the documentation for the
```
Comment (hilakashai) [inline doc/md/guides/orms/doctrine.mdx:340]:
```suggestion
migrations against your production database, read the documentation about the
```

**Review by masseelch — APPROVED** [review]

**Pull request merged**